src/lseq/ident.rs

```diff
@@ -1,78 +1,103 @@
 /// as well as the id of the site that inserted that node. This resolves conflicts where
 /// two sites decide to pick the same child index to allocate a fresh node
 #[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Clone, Serialize, Deserialize, Hash)]
-pub struct Identifier {
-    path: Vec<(u64, u32)>,
+pub struct Identifier<A: Ident> {
+    path: Vec<(u64, A)>,
     // site_id: u64,
     // counter: u64
 }
-impl Identifier {
-    /// Get the size of an identifier
+
+impl<A: Ident> Identifier<A> {
+    /// Get the size of an identifier.
     pub fn len(&self) -> usize {
         self.path.len()
     }
+
+    /// Check if the identifier is empty.
+    pub fn is_empty(&self) -> bool {
+        self.path.is_empty()
+    }
 }
-/// Generates fresh identifiers
+
+/// A generator for fresh identifiers.
 ///
 /// These identifiers represent a path in an exponential tree. At each level of the tree the amount
 /// of children doubles. This means that we can store a lot of nodes in a very low height!
 ///
-/// The LSeq tree also has an additional restriction: at each level the minimum and maximum nodes
+/// The [`crate::LSeq`] tree also has an additional restriction: at each level the minimum and maximum nodes
 /// cannot be chosen when allocating fresh nodes. This is to ensure there is always a free node
 /// that can be used to create a lower level.
 //#[derive(Serialize, Deserialize)]
-pub struct IdentGen {
+pub struct IdentGen<A: Ident> {
     initial_base_bits: u32,
     strategy_vec: BitVec,
     /// Site id of this tree
-    pub site_id: u32,
+    pub site_id: A,
     // clock: u64
 }
 
-impl IdentGen {
-    /// Create a fresh tree with 0 node
-    pub fn new(site_id: u32) -> IdentGen {
+impl<A: Ident> IdentGen<A> {
+    /// Create a fresh tree with 0 node.
+    pub fn new(site_id: A) -> Self {
         Self::new_with_args(INITIAL_BASE, site_id)
     }
 
-    /// Create a tree with a custom initial size
-    pub fn new_with_args(base: u32, site_id: u32) -> IdentGen {
-        IdentGen { initial_base_bits: base, strategy_vec: BitVec::new(), site_id }
+    /// Create a tree with a custom initial size.
+    pub fn new_with_args(base: u32, site_id: A) -> Self {
+        IdentGen {
+            initial_base_bits: base,
+            strategy_vec: BitVec::new(),
+            site_id,
+        }
     }
 
-    /// The smallest possible node in a tree
-    pub fn lower(&self) -> Identifier {
-        Identifier { path: vec![(0, 0)] }
+    /// The smallest possible node in a tree.
+    pub fn lower(&self) -> Identifier<A> {
+        Identifier {
+            path: vec![(0, A::default())],
+        }
     }
 
-    /// The absolute largest possible node
-    pub fn upper(&self) -> Identifier {
-        Identifier { path: vec![(2u64.pow(self.initial_base_bits) - 1, 0)] }
+    /// The absolute largest possible node.
+    pub fn upper(&self) -> Identifier<A> {
+        Identifier {
+            path: vec![(2u64.pow(self.initial_base_bits) - 1, A::default())],
```

Comment:
by using `u64` the `LSEQ` is actually restricted to `63 - initial_base_bits` levels. That invariant should be checked where appropriate to exit early.

Reply:
I don't understand this :) I'm not familiar with the paper and implementation, unfortunately. Let me know if I should read up on it to get a better understanding :v:

Reply:
It's mostly a comment for myself :) in theory if `initial_base_bits > 63` this would overflow. That's also important for the function `width_at(depth)` if `depth + initial_base_bits > 63` then it will overflow.